From e260e42f0d709ce1fed9dc21805aada343ee12bc Mon Sep 17 00:00:00 2001
From: "Langer, Thomas" <thomas.langer@intel.com>
Date: Fri, 10 Jan 2020 19:45:28 +0100
Subject: [PATCH] PONRTSYS-6161: Include the appended DTB in relocation

Without this move, the original location of the DTB was reserved
in the bootmem table. This was not a problem when relocating from
low to high address, as this block was still available when decoding
the DTB and freed up later. When relocating from high to low address,
this reserved block was still in place and prevented the allocation of
bigger contiguous memory.
This change ensures the DTB is copied and used from the new location.
---
 arch/mips/kernel/relocate.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/arch/mips/kernel/relocate.c b/arch/mips/kernel/relocate.c
index c54f41e960d6..99dd3b08f331 100644
--- a/arch/mips/kernel/relocate.c
+++ b/arch/mips/kernel/relocate.c
@@ -358,6 +358,28 @@ void *__init relocate_kernel(void)
 		if (res < 0)
 			goto out;
 
+#ifdef CONFIG_USE_OF
+#ifdef CONFIG_MIPS_RAW_APPENDED_DTB
+		/*
+		 * The appended DTB is not copied above, as it is located after
+		 * "_relocation_start" (see vmlinux.lds.S)
+		 * Copy the dtb block (size as specified in lds) and update the
+		 * pointer "fw_passed_dtb". The copy of .bss will update the
+		 * variable at the new location later.
+		 */
+		memcpy((void *)fw_passed_dtb + offset, (void *)fw_passed_dtb,
+			0x100000);
+		fw_passed_dtb += offset;
+
+		/*
+		 * Reset the cached pointer from fdt to ensure the new address
+		 * is retrieved again, instead of continuing with the original
+		 * address.
+		 */
+		initial_boot_params = NULL;
+#endif
+#endif
+
 		/*
 		 * The original .bss has already been cleared, and
 		 * some variables such as command line parameters
